python_on_whales/components/system/cli_wrapper.py

```diff
@@ -57,9 +61,39 @@ def disk_free(self) -> DiskFreeResult:
         full_cmd = self.docker_cmd + ["system", "df", "--format", "{{json .}}"]
         return DiskFreeResult(run(full_cmd))
 
-    def events(self):
-        """Not yet implemented"""
-        raise NotImplementedError
+    def events(self, filters: List[str]) -> List[DockerEvent]:
```

Comment:
```suggestion
    def events(self, until: Union[None, datetime.datetime, datetime.timedelta], filters: Dict[str, str]) -> Iterator[DockerEvent]:
```
The main usage of `docker system events` is streaming, as such Iterator will only work in this case. In the case where we have `--until=...` where the listening loop will exit at some point, an iterator still works. 

If the user wants a list, the user should do `list(docker.system.events(...)`.

I believe it might be a bit too hard/confusing for users to only support the case where the function will end at some point. As such, I would prefer that we implement directly the general case (iterator of infinite event) and the user will be able to pass the `until` argument if he/she wants the iterator to stop at some point.

I will guide you to implement this if you're willing to work on it :)

For the filters, I would like to use a dict to be consitent with the other functions. You can take as reference the implementation of `docker.container.prune(...)`.

For the "until" argument, we have something similar in `docker.container.logs()`, we use a datetime or a timedelta. Please do the same here. You can copy the code as needed.

We should return an iterator, as such we need to use the same code as in functions like `docker.run(stream=True)` or `docker.container.logs()`. You should use the function ` stream_stdout_and_stderr` instead of `run`.

As you get the stdout out of `stream_stdout_and_stderr`, you will be able to convert it into a `DockerEvent` and then `yield` it to the user.

Reply:
Thanks for the feedback! I'll work on a revision.

Reply:
Pushed some changes to match what you suggested.